src/channel-qvm.lisp

```diff
@@ -0,0 +1,204 @@
+;;;; channel-qvm.lisp
+;;;;
+;;;; Authors: Sophia Ponte, Nikolas Tezak, Erik Davis
+
+(in-package #:qvm)
+
+;;; Implements a QVM that can interpret a rule-based noise model and
+;;; apply kraus operators to it's state.
+
+(defgeneric apply-all-kraus-maps (qvm instr kraus-maps)
+  (:documentation "Applies every kraus-map in the list KRAUS-MAPS to the state of the system."))
+
+
+(defgeneric apply-kraus-map (qvm instr kraus-ops)
+  (:documentation "Applies noise from a kraus map to the system. Randomly select a kraus operator from the kraus map KRAUS-OPS to apply to the state of the system."))
+
+
+;;; The CHANNEL-QVM requires an explicitly defined NOISE-MODEL. This
+;;; NOISE-MODEL should be a complete specification of the different
+;;; types of noise that should be applied throughout the execution of
+;;; a QUIL program, as well as when in the program's execution the
+;;; noise should be applied. These decisions are made during
+;;; TRANSITION. At each new instruction, the CHANNEL-QVM checks to see
+;;; if any NOISE-RULES are matched by the instuction, and if so, the
+;;; CHANNEL-QVM applied the corresponding noise.
+(defclass channel-qvm (pure-state-qvm)
+  ((original-amplitude-pointer
+    :reader original-amplitude-pointer)
+   (trial-amplitudes
+    :accessor %trial-amplitudes)
+   (noise-model
+    :initarg :noise-model
+    :accessor noise-model
+    :initform nil))
```

Comment:
Suggestion: move this into `:default-initargs` for consistency with the rest of our code.